### PR TITLE
[469348]: Add magic functions [step 2]

### PR DESCRIPTION
**Pull request by danzatt**
I have rebased my commit on top of the meanwhile updated code.

### CODE REVIEW

**Comment by danzatt** [comment]
Done.

**Comment by danzatt** [comment]
608ea3183c131d73991c25994f36c0ae6dc13466 - Add magic functions' implementation, tests and documentation (rebase and fix conflicts)
049dfccb68a552c92dbf0aafa27a1ca7297d0cce - Add size checks, minor fixes and tests
6b177d500efc3db20c9aed1e4975fcebfca9a260 - Add magic functions' implementation, tests and documentation (squashed)

**Review by ph4r05 — CHANGES_REQUESTED** [review]
0 tests is run in Travis. Fix required.

**Comment by ph4r05** [comment]
As seen here: https://travis-ci.com/crocs-muni/open-source-development-course-hw02-1/jobs/282060671

Hints:

- check what .travis.yml does precisely, execute same commands. Are tests executed?
- inspect PR fixing the problem: crocs-muni/open-source-development-course-hw02-2#1

**Comment by danzatt** [comment]
Should be fixed now. A new commit list
608ea3183c131d73991c25994f36c0ae6dc13466 - Add magic functions' implementation, tests and documentation (rebase and fix conflicts)
049dfccb68a552c92dbf0aafa27a1ca7297d0cce - Add size checks, minor fixes and tests
f093bea16cbfb3bf4697812fa45b593e8e542de4 - Fix Travis tests
a0414554b68a24472f85f289fba2d7a7c0a08829 - Fix missing return and compare based on the vector length
cb201c401bb3b2c54139974c41a5921b3e984da4 - Fix tests; use the correct operator

4585c7a1a6670fa0e6de4afda719c8d03db8ed30 - Add magic functions' implementation, tests and documentation (squashed again)

**Review by ph4r05 — APPROVED** [review]

File: ossdev/__init__.py [hunk 1]
```diff
@@ -21,51 +25,183 @@ def set(self, arr):
...
-        # TODO: implement
-        return None
+        self.d[key] = value
 
     def __cmp__(self, other):
-        # TODO: implement, -1 if self < other, 0 if self == other, 1 if self > other
-        return -1
+        return other is not None and self.length() - other.length()
+
+    def __eq__(self, other):
```
Comment (ph4r05) [inline ossdev/__init__.py:56]: Why is the comparison done in this way? `__cmp__` substitutes all other magic comparison methods.
Moreover, it is quite similar to this https://github.com/crocs-muni/open-source-development-course-hw02-1/pull/4/files  why?

Comment (danzatt) [reply ossdev/__init__.py:56]: At first I tried implementing only __cmp__ but when I tested comparison it gave me an error
```
python3 -i ossdev/__init__.py
>>> a = Vector()
>>> b = Vector()
>>> a < b 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'Vector' and 'Vector'
```
Then I tried running the same code with python 2, which worked.
```
$ python2 -i ossdev/__init__.py
>>> a = Vector()
>>> b = Vector()
>>> a < b 
False
```
I searched the documentation you linked us to (https://rszalski.github.io/magicmethods/) for `__cmp__` and I found:
> # Appendix 2: Changes in Python 3
> ...
> - `__cmp__` is gone due to redundancy with other magic methods

So I thought that it is necessary to implement the other functions (as the library info stated required version is 3).

Comment (ph4r05) [reply ossdev/__init__.py:56]: Ok cool. Next time you can use https://docs.python.org/3/library/functools.html#functools.total_ordering to simplify the task.

Still, comparing by vector length only is not enough. However, we wont fix it now.
Consider this closed. Thanks.

**Comment by ph4r05** [comment]
Approved, HW done, closing

**Pull request closed without merging**